phoenix-core-client/src/main/java/org/apache/phoenix/iterate/OffsetResultIterator.java

```diff
@@ -32,32 +34,49 @@
  */
 public class OffsetResultIterator extends DelegateResultIterator {
     private int rowCount;
-    private int offset;
+    private final int offset;
+    private Tuple lastScannedTuple;
     private long pageSizeMs = Long.MAX_VALUE;
+    private boolean isIncompatibleClient = false;
 
     public OffsetResultIterator(ResultIterator delegate, Integer offset) {
         super(delegate);
         this.offset = offset == null ? -1 : offset;
+        this.lastScannedTuple = null;
     }
 
-    public OffsetResultIterator(ResultIterator delegate, Integer offset, long pageSizeMs) {
+    public OffsetResultIterator(ResultIterator delegate, Integer offset, long pageSizeMs,
+                                boolean isIncompatibleClient) {
         this(delegate, offset);
         this.pageSizeMs = pageSizeMs;
+        this.isIncompatibleClient = isIncompatibleClient;
     }
+
     @Override
     public Tuple next() throws SQLException {
+        long startTime = EnvironmentEdgeManager.currentTimeMillis();
         while (rowCount < offset) {
             Tuple tuple = super.next();
-            if (tuple == null) { return null; }
+            if (tuple == null) {
+                return null;
+            }
             if (tuple.size() == 0 || isDummy(tuple)) {
                 // while rowCount < offset absorb the dummy and call next on the underlying scanner
                 continue;
             }
             rowCount++;
-            // no page timeout check at this level because we cannot correctly resume
-            // scans for OFFSET queries until the offset is reached
+            lastScannedTuple = tuple;
+            if (!isIncompatibleClient) {
+                if (EnvironmentEdgeManager.currentTimeMillis() - startTime >= pageSizeMs) {
```

Comment:
Since we cannot handle paging before skipping offset number of rows, we should not time out here either. We should also lock the region to prevent region movement.

Reply:
We have region lock in `getOffsetScanner()` so this whole operation is under region lock.
However, the reason why any lower/upper level dummy is getting ignored as of now is because `getOffsetKvWithLastScannedRowKey()` returns the last scanned rowkey and discard dummy.

So in reality, it's true that we can remove this logic of timeout here, but even if we don't, `RegionScannerResultIterator` will continue `scanner#nextRaw` as dummy is discarded by `getOffsetKvWithLastScannedRowKey()`.

However, now i see that we should not timeout here because even if dummy is discarded, when the offset scanner returns the result, region lock is lost and hence before `RegionScannerResultIterator` performs next `scanner#nextRaw` call, region can potentially move. Let me revisit this.

Reply:
Another issue with offset is that when the region moves the hbase client will adjust the scan start key but it has no idea about the `OFFSET` attribute so the new scan will start with the correct key but wrong offset. @virajjasani are you handling this case ?

Reply:
Hmm that's also good point, let me check what happens to `_RowOffset`, i think scan attributes are preserved because when scanner is reset, on the same scan object, start rowkey is changed so the scan attributes should remain the same but let me spend sometime to confirm this.

Reply:
@virajjasani Did you take a look at this

Reply:
yes, the scan attributes remain unaffected because hbase client keeps the same scan object and it just updates start rowkey:

```
    // Else, its signal from depths of ScannerCallable that we need to reset the scanner.
    if (this.lastResult != null) {
      // The region has moved. We need to open a brand new scanner at the new location.
      // Reset the startRow to the row we've seen last so that the new scanner starts at
      // the correct row. Otherwise we may see previously returned rows again.
      // If the lastRow is not partial, then we should start from the next row. As now we can
      // exclude the start row, the logic here is the same for both normal scan and reversed scan.
      // If lastResult is partial then include it, otherwise exclude it.
      scan.withStartRow(lastResult.getRow(), lastResult.mayHaveMoreCellsInRow());
    }
```

So, we are good from scan attributes viewpoint. I will address the remaining comments, including offset queries today.

Reply:
But the offset should change because some rows have already been scanned so this won't work right @virajjasani ?

Reply:
I am not able to reproduce any issue with this.

OffsetScanner will return valid tuple with cell CF:CQ as `f_offset:c_offset`, from which the cell value will be used by SerialIterators to set offset on next scan:

```
            while (index < scans.size()) {
                Scan currentScan = scans.get(index++);
                if (remainingOffset != null) {
// <======= set offset value
                    currentScan.setAttribute(BaseScannerRegionObserverConstants.SCAN_OFFSET, PInteger.INSTANCE.toBytes(remainingOffset));
                }
                ScanMetricsHolder scanMetricsHolder =
                        ScanMetricsHolder.getInstance(readMetrics, tableName, currentScan,
                            context.getConnection().getLogLevel());
                TableResultIterator itr =
                        new TableResultIterator(mutationState, currentScan, scanMetricsHolder,
                                renewLeaseThreshold, plan, scanGrouper, caches, maxQueryEndTime);
                PeekingResultIterator peekingItr = iteratorFactory.newIterator(context, itr, currentScan, tableName, plan);
                Tuple tuple;
                if ((tuple = peekingItr.peek()) == null) {
                    peekingItr.close();
                    continue;
// <======== retrieve offset value from server
                } else if ((remainingOffset = QueryUtil.getRemainingOffset(tuple)) != null) {
                    peekingItr.next();
                    peekingItr.close();
                    continue;
                }
                context.getConnection().addIteratorForLeaseRenewal(itr);
                return peekingItr;
            }
```

Hence, depending on the cell returned by server, next scan's offset attribute will be set and accordingly new scanner will take that value.

Basically, i will have to remove page timeout logic from offset because of it's tricky to manage returning dummy, and if we do that, we will anyways ensure that server is able to return valid offset value to client and client will check whether more serial iteration needs to happen depending on how many rows were scanned.

The logic remains same even if region moves after returning valid row. The problem happens only when region moves after dummy is returned, so we will remove page timeout for offset scanner.

Reply:
That is why we had the original comment 
`// no page timeout check at this level because we cannot correctly resume
          scans for OFFSET queries until the offset is reached`
 I was under the impression that you were fixing that issue too

Reply:
I fixed handling of dummy for offset, the tests also fail if we don't return right rowkey for offset dummy.
For every dummy cell that we send from server, we need to send rowkey as scan start rowkey (inclusive or exclusive depending on whether it was set as inclusive).

Reply:
Fixing page timeout for offset https://github.com/apache/phoenix/pull/1736/commits/5677efde8fc6d6046d5a8fd7a81d4e0c0ade2b45